rust/src/enhancers/grammar.rs

```diff
@@ -1,179 +1,350 @@
+use std::borrow::Cow;
+
+use anyhow::{anyhow, Context};
 
 use super::actions::{Action, FlagAction, FlagActionType, Range, VarAction};
 use super::matchers::{FrameOffset, Matcher};
 use super::rules::Rule;
-use super::Cache;
+use super::RegexCache;
+
+/// Strips the prefix `pat` from `input` and returns the rest.
+///
+/// Returns an error if `input` doesn't start with `pat.`
+fn expect<'a>(input: &'a str, pat: &str) -> anyhow::Result<&'a str> {
+    input
+        .strip_prefix(pat)
+        .ok_or_else(|| anyhow!("at `{input}`: expected `{pat}`"))
+}
 
-fn ident(input: &str) -> IResult<&str, &str> {
-    take_while1(|c: char| c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | '-'))(input)
+/// Parses a string into a bool.
+///
+/// `"1"`, `"yes"`, and `"true"` parse to `true`,
+/// `"0"`, `"no"`, and `"false"` parse to `false`,
+/// and anything else is an error.
+fn bool(input: &str) -> anyhow::Result<bool> {
+    match input {
+        "1" | "yes" | "true" => Ok(true),
+        "0" | "no" | "false" => Ok(false),
+        _ => anyhow::bail!("at `{input}`: invalid boolean value"),
+    }
 }
 
-fn rule_bool(input: &str) -> IResult<&str, bool> {
-    alt((
-        value(true, alt((tag("1"), tag("yes"), tag("true")))),
-        value(false, alt((tag("0"), tag("no"), tag("false")))),
-    ))(input)
+/// Parses an "identifier" and returns it together with the rest of the input.
+///
+/// An "identifier" is defined by the regex `[a-zA-Z0-9_.-]+`.
+fn ident(input: &str) -> anyhow::Result<(&str, &str)> {
+    let Some(end) =
+        input.find(|c: char| !(c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | '-')))
+    else {
+        return Ok((input, ""));
+    };
+
+    if end == 0 {
+        anyhow::bail!("at `{input}`: invalid identifier");
+    }
+
+    Ok(input.split_at(end))
 }
 
-fn rule_number(input: &str) -> IResult<&str, usize> {
-    map(take_while1(|c: char| c.is_ascii_digit()), |n: &str| {
-        n.parse().unwrap()
-    })(input)
+/// Parses an "argument", i.e., the right-hand side of a matcher definition, and returns it together
+/// with the rest of the input.
+///
+/// An "argument" is either a sequence of non-whitespace ASCII characters or any sequence of
+/// non-`"` characters enclosed in `""`.
+///
+/// Escaped characters in the argument are unescaped.
+fn argument(input: &str) -> anyhow::Result<(Cow<str>, &str)> {
+    let (result, rest) = if let Some(rest) = input.strip_prefix('"') {
+        let end = rest
+            .find('"')
+            .ok_or_else(|| anyhow!("at `{input}`: unclosed `\"`"))?;
+        let result = &rest[..end];
+        let rest = &rest[end + 1..];
+        (result, rest)
+    } else {
+        match input.find(|c: char| c.is_ascii_whitespace()) {
+            None => (input, ""),
+            Some(end) => input.split_at(end),
+        }
+    };
+
+    // TODO: support even more escapes
+    let unescaped = if result.contains("\\\\") {
+        result.replace("\\\\", "\\").into()
```

Comment:
Nit: The current solution iterates through the `result` input and later returns a boolean. For replace, it does the same exact thing. But ideally, we can just store the index of the matched substring and replace starting from there to reduce the search space of this operations.

Reply:
Hm, I'm not sure this plays well with the way `replace` works.
```rust
 if let idx = result.find("\\\\") {
    let replaced = &result[idx..].replace("\\\\", "\\");
    // what about `result[..idx]`?
}
```
I think it would necessitate copying string data back and forth …